src/classes/BDI_ObjectMappingLogic.cls

```diff
@@ -0,0 +1,124 @@
+/*
+    Copyright (c) 2014 Salesforce.org
+    All rights reserved.
+    
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+    
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+ 
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Batch Data Import
+* @group-content ../../ApexDocContent/BatchDataImport.htm
+* @description Virtual class containing the standard logic for mapping fields between source and target objects.  Can be overriden by clients
+* who need custom mapping logic.
+*/
+
+public with sharing virtual class BDI_ObjectMappingLogic {
+
+    public virtual BDI_ObjectWrapper[] populateObjects(BDI_ObjectWrapper[] objWraps) {
+        if (objWraps == null) {
+            return objWraps;
+        }
+
+        Map<String, Schema.DescribeFieldResult> sourceFieldDescribeMap = UTIL_Describe.getAllFieldsDescribe(UTIL_Namespace.adjustNPSPNamespace('npsp__DataImport__c'));
+        
+        for (BDI_ObjectWrapper objWrap : objWraps) {
+            Map<String, Schema.DescribeFieldResult> targetFieldDescribeMap = UTIL_Describe.getAllFieldsDescribe(objWrap.objMapping.Object_API_Name__c);
+
+            objWrap.sObj = Schema.getGlobalDescribe().get(objWrap.objMapping.Object_API_Name__c).newSObject();
+
+            // If the object already has an existing SObject Id, then set it in the sObj
+            if (objWrap.existingSObjectId != null) {
+                objWrap.sObj.put('Id', objWrap.existingSObjectId);
+            }
```

Comment:
See other comment, this is intentionally an overrideable method to increase flexibility.

Reply:
Ok, thanks.  Final question: would it make sense to put this in the else block of the `prepareObjectWrappers` method within the `BDI_AdditionalObjectService` class around [line 157](https://github.com/SalesforceFoundation/Cumulus/blob/029797fcd536f5cf1343618c8f712c6e3121d209/src/classes/BDI_AdditionalObjectService.cls#L157)? Since at that point the objWrapper has an existing SObjectId?

Reply:
I'd prefer to just keep it with all the other mapping logic inside that method in case there is some edge case reason that a client would want to do it differently.